tutorbackup/templates/backup/build/backup/backup_services.py

```diff
@@ -0,0 +1,186 @@
+
+
+ENV = os.environ
+
+DUMP_DIRECTORY = '/data'
+
+MYSQL_DUMPFILE = os.path.join(DUMP_DIRECTORY, 'mysql_dump.sql')
+MONGODB_DUMPDIR = os.path.join(DUMP_DIRECTORY, 'mongodb_dump')
+CADDY_DUMPDIR = os.path.join(DUMP_DIRECTORY, 'caddy')
+
+TARFILE = '/backup/backup.tar.xz'
+
+logger = logging.getLogger(__name__)
+
+PERFORM_CADDY_BACKUP = os.path.isdir('caddy')
+
+
+def get_size(dir_path):
+    total_size = os.path.getsize(dir_path)
+    for item in os.listdir(dir_path):
+        item_path = os.path.join(dir_path, item)
+        if os.path.isfile(item_path) and not os.path.islink(item_path):
+            total_size += os.path.getsize(item_path)
+        elif os.path.isdir(item_path):
+            total_size += get_size(item_path)
+    return total_size
+
+
+def mysqldump():
+    user = ENV['MYSQL_ROOT_USERNAME']
+    password = ENV['MYSQL_ROOT_PASSWORD']
+    host = ENV['MYSQL_HOST']
+    port = ENV['MYSQL_PORT']
+    outfile = MYSQL_DUMPFILE
+
+    logger.info(f"Dumping MySQL databases on {host}:{port} to {outfile}")
+    cmd = ("mysqldump "
+           "--all-databases --add-drop-database --routines "
+           "--events --single-transaction "
+           "--quick --quote-names --max-allowed-packet=16M "
+           f"--host={host} --port={port} "
+           f"--user={user} --password={password}")
+    with open(outfile, 'wb') as out:
+        check_call(cmd,
+                   shell=True,
+                   stdout=out,
+                   stderr=sys.stderr)
+
+    size = os.path.getsize(outfile)
+    logger.info(f"Complete. {outfile} is {size} bytes.")
+
+
+def mongodump():
+    host = ENV['MONGODB_HOST']
+    port = ENV['MONGODB_PORT']
+    outdir = MONGODB_DUMPDIR
+
+    logger.info(f"Dumping MongoDB databases on {host}:{port} to {outdir}")
+    cmd = ("mongodump "
+           f"--out={outdir} "
+           f"--host={host} "
+           f"--port={port}")
+    try:
+        cmd += (f" --username={ENV['MONGODB_USERNAME']} "
+                f"--password={ENV['MONGODB_PASSWORD']}")
+    except KeyError:
+        pass
+
+    check_call(cmd,
+               shell=True,
+               stdout=sys.stdout,
+               stderr=sys.stderr)
+
+    total_size = get_size(outdir)
+    logger.info(f"Complete. {outdir} total size {total_size} bytes.")
+
+
+def caddydump():
+    outdir = CADDY_DUMPDIR
+    logger.info(f"Copying Caddy data to {outdir}")
+    shutil.copytree('caddy', outdir)
+
+    total_size = get_size(outdir)
+    logger.info(f"Complete. {outdir} total size {total_size} bytes.")
+
+
+def archive():
+    paths = [MYSQL_DUMPFILE, MONGODB_DUMPDIR]
+    if PERFORM_CADDY_BACKUP:
+        paths.append(CADDY_DUMPDIR)
+
+    outfile = TARFILE
+
+    logger.info(f"Creating archive {outfile}")
+    with tarfile.open(outfile, "w:xz") as tar:
+        for item in paths:
+            tar.add(item)
+
+    size = os.path.getsize(outfile)
+    logger.info(f"Complete. {outfile} is {size} bytes.")
+
+
+def upload_to_s3():
+    import hashlib
```

Comment:
You probably want to put `import hashlib` at the top of the file with all the other import statements. The `s3_client` import that immediately follows is only here so that things wouldn't blow up if invoking the script without `--upload` and the S3 envars not being set.

Reply:
Ah I forgot to move that. Thanks